crates/forge-runner/src/lib.rs

```diff
@@ -0,0 +1,491 @@
+use camino::Utf8PathBuf;
+use console::style;
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use once_cell::sync::Lazy;
+use scarb_artifacts::StarknetContractArtifacts;
+use smol_str::SmolStr;
+use starknet::core::types::BlockId;
+use std::collections::HashMap;
+use std::sync::Arc;
+use test_collector::FuzzerConfig;
+use test_collector::TestCase;
+use test_collector::{ForkConfig, LinkedLibrary, RawForkParams};
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::task::JoinHandle;
+use url::Url;
+
+mod fuzzer;
+mod running;
+pub mod test_case_summary;
+pub mod test_crate_summary;
+mod test_execution_syscall_handler;
+
+pub const CACHE_DIR: &str = ".snfoundry_cache";
+
+pub static BUILTINS: Lazy<Vec<&str>> = Lazy::new(|| {
+    vec![
+        "Pedersen",
+        "RangeCheck",
+        "Bitwise",
+        "EcOp",
+        "Poseidon",
+        "SegmentArena",
+        "GasBuiltin",
+        "System",
+    ]
+});
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ForkTarget {
+    name: String,
+    params: RawForkParams,
+}
+
+impl ForkTarget {
+    #[must_use]
+    pub fn new(name: String, params: RawForkParams) -> Self {
+        Self { name, params }
+    }
+
+    #[must_use]
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    #[must_use]
+    pub fn params(&self) -> &RawForkParams {
+        &self.params
+    }
+}
+
+/// Configuration of the test runner
+#[derive(Debug, PartialEq)]
+pub struct RunnerConfig {
+    pub workspace_root: Utf8PathBuf,
+    pub exit_first: bool,
+    pub fork_targets: Vec<ForkTarget>,
+    pub fuzzer_runs: u32,
+    pub fuzzer_seed: u64,
+}
+
+impl RunnerConfig {
+    /// Creates a new `RunnerConfig` from given arguments
+    ///
+    /// # Arguments
+    ///
+    /// * `test_name_filter` - Used to filter test cases by names
+    /// * `exact_match` - Should test names match the `test_name_filter` exactly
+    /// * `exit_first` - Should runner exit after first failed test
+    #[must_use]
+    #[allow(clippy::too_many_arguments, clippy::fn_params_excessive_bools)]
+    pub fn new(
+        workspace_root: Utf8PathBuf,
+        exit_first: bool,
+        fork_targets: Vec<ForkTarget>,
+        fuzzer_runs: u32,
+        fuzzer_seed: u64,
+    ) -> Self {
+        Self {
+            workspace_root,
+            exit_first,
+            fork_targets,
+            fuzzer_runs,
+            fuzzer_seed,
+        }
+    }
+}
+
+pub struct RunnerParams {
+    corelib_path: Utf8PathBuf,
+    contracts: HashMap<String, StarknetContractArtifacts>,
+    predeployed_contracts: Utf8PathBuf,
+    environment_variables: HashMap<String, String>,
+    linked_libraries: Vec<LinkedLibrary>,
+}
+
+impl RunnerParams {
+    #[must_use]
+    pub fn linked_libraries(&self) -> &Vec<LinkedLibrary> {
+        &self.linked_libraries
+    }
+    #[must_use]
+    pub fn corelib_path(&self) -> &Utf8PathBuf {
+        &self.corelib_path
+    }
+}
+
+impl RunnerParams {
+    #[must_use]
+    pub fn new(
```

Comment:
Same as above

Reply:
I want to keep fields of this struct and other completely private/inaccessible, but we still have to allow constructing instances of it.